In [1]:
import numpy as np 
import urdf2casadi.urdfparser as u2c
from urdf2casadi.geometry import plucker
from urdf_parser_py.urdf import URDF, Pose
import PyKDL as kdl
import kdl_parser_py.urdf as kdlurdf
from timeit import Timer, timeit, repeat
import rbdl
import pybullet as pb


In [2]:
def median(lst):
    n = len(lst)
    if n < 1:
            return None
    if n % 2 == 1:
            return sorted(lst)[n//2]
    else:
            return sum(sorted(lst)[n//2-1:n//2+1])/2.0
        
def average(lst): 
    return sum(lst) / len(lst) 


def c_kdl_func():
    for j in range(njoints):
        q_kdl[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_kdl[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2

    kdl.ChainDynParam(kdlmodel, gravity_kdl).JntToCoriolis(q_kdl, qdot_kdl, g_kdl)


def c_u2c_func():   
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2

    c_u2c(q_none, qdot_none)
    
def c_pb_func():
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    c_pb = pb.calculateInverseDynamics(pbmodel, q_none, qdot_none, qddot)
    
    
def c_rbdl_func():
    for j in range(njoints):
        q_np[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_np[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    rbdl.NonlinearEffects(rbdlmodel, q_np, qdot_np, c_rbdl)


def g_kdl_func():
    for j in range(njoints):
        q_kdl[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    kdl.ChainDynParam(kdlmodel, gravity_kdl).JntToGravity(q_kdl, g_kdl)


def g_u2c_func():   
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    g_u2c(q_none)
    
def g_pb_func():
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    g_pb = pb.calculateInverseDynamics(pbmodel, q_none, qdot, qddot)
    
def g_rbdl_func():
    for j in range(njoints):
        q_np[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    rbdl.InverseDynamics(rbdlmodel, q_np, qdot_np, qddot_np, g_rbdl)
    
    
def id_u2c_func():   
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qddot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    id_u2c(q_none, qdot_none, qddot_none)
    
    
def id_pb_func():
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qddot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2 
    id_pb = pb.calculateInverseDynamics(pbmodel, q_none, qdot_none, qddot_none)
    
    
def id_rbdl_func():
    for j in range(njoints):
        q_np[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_np[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qddot_np[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    rbdl.InverseDynamics(rbdlmodel, q_np, qdot_np, qddot_np, id_rbdl)

    
def m_kdl_func():
    for j in range(njoints):
        q_kdl[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    kdl.ChainDynParam(kdlmodel, gravity_kdl).JntToMass(q_kdl, m_kdl)


def m_u2c_func():   
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    m_u2c(q_none)
    
def m_pb_func():
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    m = pb.calculateMassMatrix(pbmodel, q_none)
    
    
def m_rbdl_func():
    for j in range(njoints):
        q_np[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    rbdl.CompositeRigidBodyAlgorithm(rbdlmodel, q_np, m_rbdl)
    
    
def aba_u2c_func():   
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        tau_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    fd_aba_u2c(q_none, qdot_none, tau_none)
    
    
def crba_u2c_func():
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        tau_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    fd_crba_u2c(q_none, qdot_none, tau_none)
    
    
def fd_rbdl_func():
    for j in range(njoints):
        q_np[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_np[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        tau_np[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    rbdl.ForwardDynamics(rbdlmodel, q_np, qdot_np, tau_np, qddot_np)
    

In [3]:
#pbmodel = pb.loadURDF("pantilt.urdf")
sim = pb.connect(pb.DIRECT)
#pbmodel = pb.loadURDF("/urdf4timing/1dof.urdf")
ndofs = 10
urdf_nr = list(range(ndofs + 1))
urdf_nr.pop(0)


nitr = 1000
nrepeat = 100
    

path_to_urdf = "/home/lmjohann/urdf2casadi/examples/urdf/ur5_mod.urdf"
root = 'base_link'
tip = "tool0"

ok, kdl_tree = kdlurdf.treeFromFile(path_to_urdf)
kdlmodel = kdl_tree.getChain(root,tip)

rbdlmodel = rbdl.loadModel(path_to_urdf)
pbmodel = pb.loadURDF(path_to_urdf, useFixedBase=True, flags = pb.URDF_USE_INERTIA_FROM_FILE)
asd = u2c.URDFparser()
robot_desc = asd.from_file(path_to_urdf)

jlist, names, qmax, qmin = asd.get_joint_info(root, tip)
njoints = asd.get_n_joints(root, tip)

gravity_kdl = kdl.Vector()
gravity_kdl[2] = -9.81
gravity_u2c = [0, 0, -9.81]

#rbdl declarations
q_np = np.zeros(njoints)
qdot_np = np.zeros(njoints)
qddot_np = np.zeros(njoints)
g_rbdl = np.zeros(njoints)
c_rbdl = np.zeros(njoints)
id_rbdl = np.zeros(njoints)

#kdl declatations
q_kdl = kdl.JntArray(njoints)
qdot_kdl = kdl.JntArray(njoints)
g_kdl = kdl.JntArray(njoints)

#u2c and pybullet declarations
q_none = [None]*njoints
qdot_none = [None]*njoints
qddot_none = [None]*njoints
g_pb = [None]*njoints

qddot = []
for i in range(njoints):
    qddot.append(0.)

qdot = []
for i in range(njoints):
    qdot.append(0.)
    
    
m_kdl = kdl.JntSpaceInertiaMatrix(njoints)
m_rbdl = (rbdlmodel.q_size, rbdlmodel.q_size)
m_rbdl = np.zeros(m_rbdl)
tau_np = np.zeros(njoints)
tau_none = [None]*njoints

g_u2c = asd.get_gravity_rnea(root, tip, gravity_u2c)
c_u2c = asd.get_coriolis_rnea(root, tip)
m_u2c = asd.get_inertia_matrix_crba(root, tip)
fd_aba_u2c = asd.get_forward_dynamics_aba(root, tip, gravity_u2c)
fd_crba_u2c = asd.get_forward_dynamics_crba(root, tip, gravity_u2c)
id_u2c = asd.get_inverse_dynamics_rnea(root, tip, gravity_u2c)




The root link base_link has an inertia specified in the URDF, but KDL does not support a root link with an inertia.  As a workaround, you can add an extra dummy link to your URDF.
/home/lmjohann/urdf2casadi/examples/urdf/ur5_mod.urdf


In [4]:
timeit_g_u2c = repeat("g_u2c_func()", setup = "from __main__ import g_u2c_func", repeat = nrepeat, number = nitr)
mediantime_g_u2c  = median(timeit_g_u2c)/nitr*1000000

timeit_g_pb = repeat("g_pb_func()", setup = "from __main__ import g_pb_func", repeat = nrepeat, number = nitr)
mediantime_g_pb = median(timeit_g_pb)/nitr*1000000

timeit_g_rbdl = repeat("g_rbdl_func()", setup = "from __main__ import g_rbdl_func", repeat = nrepeat, number = nitr)
mediantime_g_rbdl = median(timeit_g_rbdl)/nitr*1000000

timeit_g_kdl = repeat("g_kdl_func()", setup = "from __main__ import g_kdl_func", repeat = nrepeat, number = nitr)
mediantime_g_kdl = median(timeit_g_kdl)/nitr*1000000


timeit_c_u2c = repeat("c_u2c_func()", setup = "from __main__ import c_u2c_func", repeat = nrepeat, number = nitr)
mediantime_c_u2c  = median(timeit_c_u2c)/nitr*1000000

timeit_c_pb = repeat("c_pb_func()", setup = "from __main__ import c_pb_func", repeat = nrepeat, number = nitr)
mediantime_c_pb = median(timeit_c_pb)/nitr*1000000

timeit_c_rbdl = repeat("c_rbdl_func()", setup = "from __main__ import c_rbdl_func", repeat = nrepeat, number = nitr)
mediantime_c_rbdl = median(timeit_c_rbdl)/nitr*1000000

timeit_c_kdl = repeat("c_kdl_func()", setup = "from __main__ import c_kdl_func", repeat = nrepeat, number = nitr)
mediantime_c_kdl = median(timeit_c_kdl)/nitr*1000000

timeit_id_u2c = repeat("id_u2c_func()", setup = "from __main__ import id_u2c_func", repeat = nrepeat, number = nitr)
mediantime_id_u2c  = median(timeit_id_u2c)/nitr*1000000

timeit_id_pb = repeat("id_pb_func()", setup = "from __main__ import id_pb_func", repeat = nrepeat, number = nitr)
mediantime_id_pb = median(timeit_id_pb)/nitr*1000000

timeit_id_rbdl = repeat("id_rbdl_func()", setup = "from __main__ import id_rbdl_func", repeat = nrepeat, number = nitr)
mediantime_id_rbdl = median(timeit_id_rbdl)/nitr*1000000


timeit_m_u2c = repeat("m_u2c_func()", setup = "from __main__ import m_u2c_func", repeat = nrepeat, number = nitr)
mediantime_m_u2c  = median(timeit_m_u2c)/nitr*1000000

timeit_m_pb = repeat("m_pb_func()", setup = "from __main__ import m_pb_func", repeat = nrepeat, number = nitr)
mediantime_m_pb = median(timeit_m_pb)/nitr*1000000

timeit_m_rbdl = repeat("m_rbdl_func()", setup = "from __main__ import m_rbdl_func", repeat = nrepeat, number = nitr)
mediantime_m_rbdl = median(timeit_m_rbdl)/nitr*1000000

timeit_m_kdl = repeat("m_kdl_func()", setup = "from __main__ import m_kdl_func", repeat = nrepeat, number = nitr)
mediantime_m_kdl = median(timeit_m_kdl)/nitr*1000000


timeit_fdaba_u2c = repeat("aba_u2c_func()", setup = "from __main__ import aba_u2c_func", repeat = nrepeat, number = nitr)
mediantime_fdaba_u2c = median(timeit_fdaba_u2c)/nitr*1000000

timeit_fdcrba_u2c = repeat("crba_u2c_func()", setup = "from __main__ import crba_u2c_func", repeat = nrepeat, number = nitr)
mediantime_fdcrba_u2c = median(timeit_fdcrba_u2c)/nitr*1000000

timeit_fd_rbdl = repeat("fd_rbdl_func()", setup = "from __main__ import fd_rbdl_func", repeat = nrepeat, number = nitr)
mediantime_fd_rbdl = median(timeit_fd_rbdl)/nitr*1000000


In [5]:
print "\nur5_g_u2c=", mediantime_g_u2c
print "\nur5_id_u2c=", mediantime_id_u2c
print "\nur5_c_u2c=", mediantime_c_u2c
print "\nur5_m_u2c=", mediantime_m_u2c

print "\nur5_g_rbdl=", mediantime_g_rbdl
print "\nur5_id_rbdl=", mediantime_id_rbdl
print "\nur5_c_rbdl=", mediantime_c_rbdl
print "\nur5_m_rbdl=", mediantime_m_rbdl

print "\nur5_g_pb=", mediantime_g_pb
print "\nur5_id_pb=", mediantime_id_pb
print "\nur5_c_pb=", mediantime_c_pb
print "\nur5_m_pb=", mediantime_m_pb

print "\nur5_g_kdl=", mediantime_g_kdl
print "\nur5_c_kdl=", mediantime_c_kdl
print "\nur5_m_kdl=", mediantime_m_kdl

print "\nfdrbdl=", mediantime_fd_rbdl
print "\nfdcrba=", mediantime_fdcrba_u2c 
print "\nfdaba=", mediantime_fdaba_u2c


ur5_g_u2c= 15.4145956039

ur5_id_u2c= 36.2019538879

ur5_c_u2c= 26.2235403061

ur5_m_u2c= 16.4244174957

ur5_g_rbdl= 6.11400604248

ur5_id_rbdl= 11.7824077606

ur5_c_rbdl= 8.58449935913

ur5_m_rbdl= 5.37157058716

ur5_g_pb= 5.01108169556

ur5_id_pb= 10.0389719009

ur5_c_pb= 7.59303569794

ur5_m_pb= 5.46944141388

ur5_g_kdl= 12.6124620438

ur5_c_kdl= 15.8319473267

ur5_m_kdl= 13.282418251

fdrbdl= 12.9934549332

fdcrba= 37.8240346909

fdaba= 37.0440483093


In [6]:

ur5_g_u2c= 15.758395195

ur5_id_u2c= 36.719083786

ur5_C_u2c= 26.3614654541

ur5_M_u2c= 16.0526037216

fdcrba= 37.8240346909

fdaba= 37.4803543091




ur5_g_rbdl= 6.27994537354

ur5_id_rbdl= 12.0129585266

ur5_C_rbdl= 8.66198539734

ur5_M_rbdl= 5.52153587341

ur5_g_pb= 5.00798225403

ur5_id_pb= 10.1070404053

ur5_C_pb= 7.7269077301

ur5_M_pb= 5.52296638489

ur5_g_kdl= 12.5441551208

ur5_C_kdl= 15.8885717392

ur5_M_kdl= 13.5569572449

fdrbdl= 13.2150650024


# Derivatives

## q

In [6]:
import casadi as cs

qs =cs.SX.sym("qs",njoints)
qsdot =cs.SX.sym("qsdot",njoints)
qsddot =cs.SX.sym("qsddot",njoints)
taus =cs.SX.sym("taus",njoints)

c_dq = cs.jacobian(c_u2c(qs, qsdot), qs)
g_dq = cs.jacobian(g_u2c(qs), qs)
m_dq = cs.jacobian(m_u2c(qs), qs)
id_dq = cs.jacobian(id_u2c(qs, qsdot, qsddot), qs)
fd_aba_dq = cs.jacobian(fd_aba_u2c(qs, qsdot, taus), qs)
fd_crba_dq = cs.jacobian(fd_crba_u2c(qs, qsdot, taus), qs)


g_dq = cs.Function("g_dq", [qs], [g_dq], {"jit": True, "jit_options":{"flags":"-Ofast"}})
c_dq = cs.Function("C_dq", [qs, qsdot], [c_dq], {"jit": True, "jit_options":{"flags":"-Ofast"}})
m_dq = cs.Function("M_dq", [qs], [m_dq], {"jit": True, "jit_options":{"flags":"-Ofast"}})
id_dq = cs.Function("id_dq", [qs, qsdot, qsddot], [id_dq], {"jit": True, "jit_options":{"flags":"-Ofast"}})
fd_aba_dq = cs.Function("fd_aba_dq", [qs, qsdot, taus], [fd_aba_dq], {"jit": True, "jit_options":{"flags":"-Ofast"}})
fd_crba_dq = cs.Function("fd_crba_dq", [qs, qsdot, taus], [fd_crba_dq], {"jit": True, "jit_options":{"flags":"-Ofast"}})

In [14]:
def fdcrba_dq_func():
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        tau_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    fd_crba_dq(q_none, qdot_none, tau_none)

def fdaba_dq_func():
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        tau_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    fd_aba_dq(q_none, qdot_none, tau_none)
    
    
def m_dq_func():   
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2   
    m_dq(q_none)
    
def g_dq_func():   
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    g_dq(q_none)


def c_dq_func():   
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2

    c_dq(q_none, qdot_none)
    
def id_dq_func():   
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qddot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2

    id_dq(q_none, qdot_none, qddot_none)

def aba_dq_func():   
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        tau_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2

    fd_aba_dq(q_none, qdot_none, tau_none)
    
def crba_dq_func():
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        tau_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    
    fd_crba_dq(q_none, qdot_none, tau_none)

In [17]:
#timeit_idj = repeat("id_dq_func()", setup = "from __main__ import id_dq_func", repeat = nrepeat, number = nitr)
mediantime_idj = median(timeit_idj)/nitr*1000000

#timeit_gj = repeat("g_dq_func()", setup = "from __main__ import g_dq_func", repeat = nrepeat, number = nitr)
mediantime_gj = median(timeit_gj)/nitr*1000000

#timeit_cj = repeat("c_dq_func()", setup = "from __main__ import c_dq_func", repeat = nrepeat, number = nitr)
mediantime_cj = median(timeit_cj)/nitr*1000000

#timeit_mj = repeat("m_dq_func()", setup = "from __main__ import m_dq_func", repeat = nrepeat, number = nitr)
mediantime_mj = median(timeit_mj)/nitr*1000000

timeit_fdabaj = repeat("aba_dq_func()", setup = "from __main__ import aba_dq_func", repeat = nrepeat, number = nitr)
mediantime_fd_abaj = median(timeit_fdabaj)/nitr*1000000

timeit_fdcrbaj = repeat("crba_dq_func()", setup = "from __main__ import crba_dq_func", repeat = nrepeat, number = nitr)
mediantime_fd_crbaj = median(timeit_fdcrbaj)/nitr*1000000

   
print "\nur5_id_d=", mediantime_idj
print "\nur5_g_d=", mediantime_gj
print "\nur5_c_d=", mediantime_cj
print "\nur5_m_d=", mediantime_mj
print "\nur5_fdaba_d=", mediantime_fd_abaj
print "\nur5_fdcrba_d=", mediantime_fd_crbaj


ur5_id_d= 37.9405021667

ur5_g_d= 15.5701637268

ur5_c_d= 27.195930481

ur5_m_d= 17.228603363

ur5_fdaba_d= 50.4510402679

ur5_fdcrba_d= 46.3529825211




ur5_id_d= 38.9980077744

ur5_g_d= 15.8679485321

ur5_C_d= 27.7240276337

ur5_M_d= 17.8405046463

ur5_fdaba_d= 46.168088913

ur5_fdcrba_d= 46.4965105057



ur5_fdaba_d= 50.4510402679

ur5_fdcrba_d= 46.3529825211


ur5_id_d= 36.6973876953

ur5_g_d= 16.0750150681

ur5_C_d= 28.6685228348

ur5_M_d= 17.9800987244

ur5_fdaba_d= 37.5105142593

ur5_fdcrba_d= 37.5329256058



ur5_id_d= 37.966966629

ur5_g_d= 15.5701637268

ur5_c_d= 27.4249315262

ur5_m_d= 17.228603363

ur5_fdaba_d= 49.1979122162

ur5_fdcrba_d= 49.2479801178


ur5_fdaba_d= 49.1869449615

ur5_fdcrba_d= 45.9444522858


ur5_fdaba_d= 48.945069313

ur5_fdcrba_d= 46.1720228195

## qdot

In [9]:
c_dqdot = cs.jacobian(c_u2c(qs, qsdot), qs)
id_dqdot = cs.jacobian(id_u2c(qs, qsdot, qsddot), qsdot)
fd_aba_dqdot = cs.jacobian(fd_aba_u2c(qs, qsdot, taus), qsdot)
fd_crba_dqdot = cs.jacobian(fd_crba_u2c(qs, qsdot, taus), qsdot)

c_dqdot = cs.Function("c_dq", [qs, qsdot], [c_dqdot], {"jit": True, "jit_options":{"flags":"-Ofast"}})
id_dqdot = cs.Function("id_dq", [qs, qsdot, qsddot], [id_dqdot], {"jit": True, "jit_options":{"flags":"-Ofast"}})
fd_aba_dqdot = cs.Function("fd_aba_dq", [qs, qsdot, taus], [fd_aba_dqdot], {"jit": True, "jit_options":{"flags":"-Ofast"}})
fd_crba_dqdot = cs.Function("fd_crba_dq", [qs, qsdot, taus], [fd_crba_dqdot], {"jit": True, "jit_options":{"flags":"-Ofast"}})

In [12]:


def c_dqdot_func():   
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2

    c_dqdot(q_none, qdot_none)
    
def id_dqdot_func():   
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qddot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2

    id_dqdot(q_none, qdot_none, qddot_none)

def aba_dqdot_func():   
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        tau_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2

    fd_aba_dqdot(q_none, qdot_none, tau_none)
    
def crba_dqdot_func():
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        tau_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    
    fd_crba_dqdot(q_none, qdot_none, tau_none)

In [20]:
#timeit_idj = repeat("id_dqdot_func()", setup = "from __main__ import id_dqdot_func", repeat = nrepeat, number = nitr)
mediantime_idj = median(timeit_idj)/nitr*1000000

#timeit_cj = repeat("c_dqdot_func()", setup = "from __main__ import c_dqdot_func", repeat = nrepeat, number = nitr)
mediantime_cj = median(timeit_cj)/nitr*1000000


timeit_fdabaj = repeat("aba_dqdot_func()", setup = "from __main__ import aba_dqdot_func", repeat = nrepeat, number = nitr)
mediantime_fd_abaj = median(timeit_fdabaj)/nitr*1000000

timeit_fdcrbaj = repeat("crba_dqdot_func()", setup = "from __main__ import crba_dqdot_func", repeat = nrepeat, number = nitr)
mediantime_fd_crbaj = median(timeit_fdcrbaj)/nitr*1000000

   
print "\nur5_id_d=", mediantime_idj
print "\nur5_C_d=", mediantime_cj
print "\nur5_fdaba_d=", mediantime_fd_abaj
print "\nur5_fdcrba_d=", mediantime_fd_crbaj


ur5_id_d= 37.9405021667

ur5_C_d= 27.195930481

ur5_fdaba_d= 38.8225317001

ur5_fdcrba_d= 38.6475324631


ur5_fdaba_d= 38.7120246887

ur5_fdcrba_d= 38.6099815369



ur5_fdaba_d= 39.2190217972

ur5_fdcrba_d= 38.8514995575


ur5_id_d= 37.7110242844

ur5_C_d= 27.3430347443

ur5_fdaba_d= 41.2844419479

ur5_fdcrba_d= 38.8004779816


## qddot / tau


In [21]:


id_dqddot = cs.jacobian(id_u2c(qs, qsdot, qsddot), qsddot)
fd_aba_dtau = cs.jacobian(fd_aba_u2c(qs, qsdot, taus), taus)
fd_crba_dtau = cs.jacobian(fd_crba_u2c(qs, qsdot, taus), taus)

id_dqddot = cs.Function("id_qddot", [qs, qsdot, qsddot], [id_dqddot], {"jit": True, "jit_options":{"flags":"-Ofast"}})
fd_aba_dtau = cs.Function("fd_aba_dtau", [qs, qsdot, taus], [fd_aba_dtau], {"jit": True, "jit_options":{"flags":"-Ofast"}})
fd_crba_dtau = cs.Function("fd_crba_dtau", [qs, qsdot, taus], [fd_crba_dtau], {"jit": True, "jit_options":{"flags":"-Ofast"}})

In [25]:
def id_dqddot_func():   
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qddot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2

    id_dqddot(q_none, qdot_none, qddot_none)

def aba_dtau_func():   
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        tau_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2

    fd_aba_dtau(q_none, qdot_none, tau_none)
    
def crba_dtau_func():
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        tau_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    
    fd_crba_dtau(q_none, qdot_none, tau_none)

In [28]:
timeit_idj = repeat("id_dqddot_func()", setup = "from __main__ import id_dqddot_func", repeat = nrepeat, number = nitr)
mediantime_idj = median(timeit_idj)/nitr*1000000

timeit_fdabaj = repeat("aba_dtau_func()", setup = "from __main__ import aba_dtau_func", repeat = nrepeat, number = nitr)
mediantime_fd_abaj = median(timeit_fdabaj)/nitr*1000000

timeit_fdcrbaj = repeat("crba_dtau_func()", setup = "from __main__ import crba_dtau_func", repeat = nrepeat, number = nitr)
mediantime_fd_crbaj = median(timeit_fdcrbaj)/nitr*1000000

   
print "\nur5_id_d=", mediantime_idj
print "\nur5_fdaba_d=", mediantime_fd_abaj
print "\nur5_fdcrba_d=", mediantime_fd_crbaj


ur5_id_d= 36.7964506149

ur5_fdaba_d= 39.6649837494

ur5_fdcrba_d= 36.8354320526


In [14]:

ur5_id_d= 36.5724563599

ur5_fdaba_d= 37.5800132751

ur5_fdcrba_d= 36.8819236755




ur5_id_d= 36.7964506149

ur5_fdaba_d= 39.6649837494

ur5_fdcrba_d= 36.8354320526




## derivatives all 

In [5]:
import casadi as cs
qs =cs.SX.sym("qs",njoints)
qsdot =cs.SX.sym("qsdot",njoints)
qsddot =cs.SX.sym("qsddot",njoints)
qsdddot = cs.SX.sym("qsdddot", njoints)
taus =cs.SX.sym("taus",njoints)
tausdot =cs.SX.sym("tausdot",njoints)

c_vec = cs.vertcat(qs, qsdot)
c_dvec = cs.vertcat(qsdot, qsddot)
id_vec = cs.vertcat(qs, qsdot, qsddot)
id_dvec = cs.vertcat(qsdot, qsdot, qsdddot)
fd_vec = cs.vertcat(qs, qsdot, qsddot)
fd_dvec = cs.vertcat(qsdot, qsdot, tausdot)

dc = cs.jacobian(c_u2c(qs, qsdot), c_vec)
dc = cs.Function("dc", [qs, qsdot], [dc], {"jit": True, "jit_options":{"flags":"-Ofast"}})

dctimes = cs.jtimes(c_u2c(qs, qsdot), c_vec, c_dvec)
dctimes = cs.Function("dctimes", [qs, qsdot, qsddot], [dctimes], {"jit": True, "jit_options":{"flags":"-Ofast"}})

did = cs.jacobian(id_u2c(qs, qsdot, qsddot), id_vec)
did = cs.Function("did", [qs, qsdot, qsddot], [did], {"jit": True, "jit_options":{"flags":"-Ofast"}})

didtimes = cs.jtimes(id_u2c(qs, qsdot, qsddot), id_vec, id_dvec)
didtimes = cs.Function("didtimes", [qs, qsdot, qsddot, qsdddot], [didtimes], {"jit": True, "jit_options":{"flags":"-Ofast"}})


dfd_aba = cs.jacobian(fd_aba_u2c(qs, qsdot, taus), fd_vec)
dfd_aba = cs.Function("dfd_aba", [qs, qsdot, taus], [dfd_aba], {"jit": True, "jit_options":{"flags":"-Ofast"}})

dfdtimes_aba = cs.jtimes(fd_aba_u2c(qs, qsdot, taus), fd_vec, fd_dvec)
dfdtimes_aba = cs.Function("dfdtimes_aba", [qs, qsdot, taus, qsddot, tausdot], [dfdtimes_aba], {"jit": True, "jit_options":{"flags":"-Ofast"}})

dfd_crba = cs.jacobian(fd_crba_u2c(qs, qsdot, taus), fd_vec)
dfd_crba = cs.Function("dfd_crba", [qs, qsdot, taus], [dfd_crba], {"jit": True, "jit_options":{"flags":"-Ofast"}})

dfdtimes_crba = cs.jtimes(fd_crba_u2c(qs, qsdot, taus), fd_vec, fd_dvec)
dfdtimes_crba = cs.Function("dfdtimes_crba", [qs, qsdot, taus, qsddot, tausdot], [dfdtimes_crba], {"jit": True, "jit_options":{"flags":"-Ofast"}})

dg = cs.jacobian(g_u2c(qs), qs)
dg = cs.Function("dg", [qs], [dg], {"jit": True, "jit_options":{"flags":"-Ofast"}})


dm = cs.jacobian(m_u2c(qs), qs)
dm = cs.Function("dm", [qs], [dm], {"jit": True, "jit_options":{"flags":"-Ofast"}})

dgtimes = cs.jtimes(g_u2c(qs), qs, qsdot)
dgtimes = cs.Function("dgtimes", [qs, qsdot], [dgtimes], {"jit": True, "jit_options":{"flags":"-Ofast"}})

dmtimes = cs.jtimes(m_u2c(qs), qs, qsdot)
dmtimes = cs.Function("dmtimes", [qs, qsdot], [dmtimes], {"jit": True, "jit_options":{"flags":"-Ofast"}})


print "\nur5_nodes_did = ", did.n_nodes()
print "\nur5_nodes_dc = ", dc.n_nodes()
print "\nur5_nodes_dm = ", dm.n_nodes()
print "\nur5_nodes_dg = ", dg.n_nodes()
print "\nur5_nodes_dfd_aba = ", dfd_aba.n_nodes()
print "\nur5_nodes_dfd_crba =", dfd_crba.n_nodes()

print "\nur5_nodes_didtimes = ", didtimes.n_nodes()
print "\nur5_nodes_dctimes = ", dctimes.n_nodes()
print "\nur5_nodes_dmtimes = ", dmtimes.n_nodes()
print "\nur5_nodes_dgtimes = ", dgtimes.n_nodes()
print "\nur5_nodes_dfdtimes_aba = ", dfdtimes_aba.n_nodes()
print "\nur5_nodes_dfdtimes_crba =", dfdtimes_crba.n_nodes()


def dfd_crba_func():
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        tau_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    dfd_crba(q_none, qdot_none, tau_none)
   
def dfdtimes_crba_func():
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        tau_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qddot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        taudot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    dfdtimes_crba(q_none, qdot_none, tau_none, qddot_none, taudot_none)

def dfd_aba_func():
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        tau_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    dfd_aba(q_none, qdot_none, tau_none)
   
def dfdtimes_aba_func():
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qddot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        taudot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    dfdtimes_aba(q_none, qdot_none, tau_none, qddot_none, taudot_none)
   
   
   
def dm_func():  
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2  
    dm(q_none)
   
def dg_func():  
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    dg(q_none)
   
   
def dmtimes_func():  
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2  
        qdot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    dmtimes(q_none, qdot_none)
   
def dgtimes_func():  
    for j in range(njoints):
        qdot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2        
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    dgtimes(q_none, qdot_none)


def dc_func():  
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2

    dc(q_none, qdot_none)
   
   
def dctimes_func():  
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qddot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2

    dctimes(q_none, qdot_none, qddot_none)
   
   
def did_func():
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qddot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    did(q_none, qdot_none, qddot_none)
   
def didtimes_func():
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qddot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdddot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    didtimes(q_none, qdot_none, qddot_none, qdddot_none)

qdddot_none = [None]*njoints
taudot_none = [None]*njoints





ur5_nodes_did =  4731

ur5_nodes_dc =  5157

ur5_nodes_dm =  4095

ur5_nodes_dg =  881

ur5_nodes_dfd_aba =  13517

ur5_nodes_dfd_crba = 17764

ur5_nodes_didtimes =  1807

ur5_nodes_dctimes =  1707

ur5_nodes_dmtimes =  2909

ur5_nodes_dgtimes =  491

ur5_nodes_dfdtimes_aba =  5828

ur5_nodes_dfdtimes_crba = 6906

ur5_did = 40.0873422623

ur5_dg = 15.3809785843

ur5_dc = 29.6190977097

ur5_dm = 17.1409845352

ur5_dfd_aba = 50.2574443817

ur5_dfd_crba = 48.3536720276


In [ ]:

ur5_nodes_did =  4731

ur5_nodes_dc =  5157

ur5_nodes_dm =  4095

ur5_nodes_dg =  881

ur5_nodes_dfd_aba =  13517

ur5_nodes_dfd_crba = 17764


ur5_nodes_didtimes =  1807

ur5_nodes_dctimes =  1707

ur5_nodes_dmtimes =  2909

ur5_nodes_dgtimes =  491

ur5_nodes_dfdtimes_aba =  5828

ur5_nodes_dfdtimes_crba = 6906

In [12]:
timeit_did = repeat("did_func()", setup = "from __main__ import did_func", repeat = nrepeat, number = nitr)
mediantime_did = median(timeit_did)/nitr*1000000

timeit_dg = repeat("dg_func()", setup = "from __main__ import dg_func", repeat = nrepeat, number = nitr)
mediantime_dg = median(timeit_dg)/nitr*1000000

timeit_dc = repeat("dc_func()", setup = "from __main__ import dc_func", repeat = nrepeat, number = nitr)
mediantime_dc = median(timeit_dc)/nitr*1000000

timeit_dm = repeat("dm_func()", setup = "from __main__ import dm_func", repeat = nrepeat, number = nitr)
mediantime_dm = median(timeit_dm)/nitr*1000000

timeit_dfd_aba = repeat("dfd_aba_func()", setup = "from __main__ import dfd_aba_func", repeat = nrepeat, number = nitr)
mediantime_dfd_aba = median(timeit_dfd_aba)/nitr*1000000

timeit_dfd_crba = repeat("dfd_crba_func()", setup = "from __main__ import dfd_crba_func", repeat = nrepeat, number = nitr)
mediantime_dfd_crba = median(timeit_dfd_crba)/nitr*1000000

   
print "\nur5_did =", mediantime_did
print "\nur5_dg =", mediantime_dg
print "\nur5_dc =", mediantime_dc
print "\nur5_dm =", mediantime_dm
print "\nur5_dfd_aba =", mediantime_dfd_aba
print "\nur5_dfd_crba =", mediantime_dfd_crba


ur5_did = 40.4714345932

ur5_dg = 15.4880285263

ur5_dc = 29.744386673

ur5_dm = 18.0056095123

ur5_dfd_aba = 52.5209903717

ur5_dfd_crba = 50.784945488



ur5_nodes_did =  4731

ur5_nodes_dc =  5157

ur5_nodes_dm =  4095

ur5_nodes_dg =  881

ur5_nodes_dfd_aba =  13517

ur5_nodes_dfd_crba = 17764

ur5_nodes_didtimes =  1807

ur5_nodes_dctimes =  1707

ur5_nodes_dmtimes =  2909

ur5_nodes_dgtimes =  491

ur5_nodes_dfdtimes_aba =  5828

ur5_nodes_dfdtimes_crba = 6906

ur5_did = 40.0873422623

ur5_dg = 15.3809785843

ur5_dc = 29.6190977097

ur5_dm = 17.1409845352

ur5_dfd_aba = 50.2574443817

ur5_dfd_crba = 48.3536720276


In [13]:


timeit_didtimes = repeat("didtimes_func()", setup = "from __main__ import didtimes_func", repeat = nrepeat, number = nitr)
mediantime_didtimes = median(timeit_didtimes)/nitr*1000000

timeit_dgtimes = repeat("dgtimes_func()", setup = "from __main__ import dgtimes_func", repeat = nrepeat, number = nitr)
mediantime_dgtimes = median(timeit_dgtimes)/nitr*1000000

timeit_dctimes = repeat("dctimes_func()", setup = "from __main__ import dctimes_func", repeat = nrepeat, number = nitr)
mediantime_dctimes = median(timeit_dctimes)/nitr*1000000

timeit_dmtimes = repeat("dmtimes_func()", setup = "from __main__ import dmtimes_func", repeat = nrepeat, number = nitr)
mediantime_dmtimes = median(timeit_dmtimes)/nitr*1000000

timeit_dfdtimes_aba = repeat("dfdtimes_aba_func()", setup = "from __main__ import dfdtimes_aba_func", repeat = nrepeat, number = nitr)
mediantime_dfdtimes_aba = median(timeit_dfdtimes_aba)/nitr*1000000

timeit_dfdtimes_crba = repeat("dfdtimes_crba_func()", setup = "from __main__ import dfdtimes_crba_func", repeat = nrepeat, number = nitr)
mediantime_dfdtimes_crba = median(timeit_dfdtimes_crba)/nitr*1000000

print "\nur5_didtimes =", mediantime_didtimes
print "\nur5_dgtimes =", mediantime_dgtimes
print "\nur5_dctimes =", mediantime_dctimes
print "\nur5_dmtimes =", mediantime_dmtimes
print "\nur5_dfdtimes_aba =", mediantime_dfdtimes_aba
print "\nur5_dfdtimes_crba =", mediantime_dfdtimes_crba


ur5_didtimes = 49.7659444809

ur5_dgtimes = 26.8129110336

ur5_dctimes = 37.9886627197

ur5_dmtimes = 30.0456285477

ur5_dfdtimes_aba = 55.1775693893

ur5_dfdtimes_crba = 69.8469877243



ur5_didtimes = 49.3636131287

ur5_dgtimes = 25.5494117737

ur5_dctimes = 37.2189283371

ur5_dmtimes = 28.6979675293

ur5_dfdtimes_aba = 54.1299581528

ur5_dfdtimes_crba = 66.9349431992



ur5_didtimes = 49.0859746933

ur5_dgtimes = 25.4321098328

ur5_dctimes = 36.9275808334

ur5_dmtimes = 28.3726453781

ur5_dfdtimes_aba = 53.7585020065

ur5_dfdtimes_crba = 67.2144889832



# Number of Operations

In [32]:
print "\nur5_nodes_id=", id_u2c.n_nodes()
print "\nur5_nodes_c=", c_u2c.n_nodes()
print "\nur5_nodes_m=", m_u2c.n_nodes()
print "\nur5_nodes_g=", g_u2c.n_nodes()
print "\nur5_nodes_fdaba", fd_aba_u2c.n_nodes()
print "\nur5_nodes_fdcrba=", fd_crba_u2c.n_nodes() 

print "\n\nur5_nodes_id_dq=", id_dq.n_nodes()
print "\nur5_nodes_c_dq=", c_dq.n_nodes()
print "\nur5_nodes_m_dq =", m_dq.n_nodes()
print "\nur5_nodes_g_dq =", g_dq.n_nodes()
print "\nur5_nodes_fdaba_dq =", fd_aba_dq.n_nodes()
print "\nur5_nodes_fdcrba_dq =", fd_crba_dq.n_nodes()     

print "\n\nur5_nodes_id_dqdot=", id_dqdot.n_nodes()
print "\nur5_nodes_c_dqdot =", c_dqdot.n_nodes()
print "\nur5_nodes_fdaba_dqdot =", fd_aba_dqdot.n_nodes()
print "\nur5_nodes_fdcrba_dqdot =", fd_crba_dqdot.n_nodes()     

print "\n\nur5_nodes_id_dqddot=", id_dqddot.n_nodes()
print "\nur5_nodes_fdaba_dtau =", fd_aba_dtau.n_nodes()
print "\nur5_nodes_fdcrba_dtau =", fd_crba_dtau.n_nodes()   


ur5_nodes_id= 696

ur5_nodes_c= 648

ur5_nodes_m= 1026

ur5_nodes_g= 199

ur5_nodes_fdaba 1948

ur5_nodes_fdcrba= 2354


ur5_nodes_id_dq= 3069

ur5_nodes_c_dq= 3006

ur5_nodes_m_dq = 4095

ur5_nodes_g_dq = 881

ur5_nodes_fdaba_dq = 10189

ur5_nodes_fdcrba_dq = 15194


ur5_nodes_id_dqdot= 2371

ur5_nodes_c_dqdot = 3006

ur5_nodes_fdaba_dqdot = 4605

ur5_nodes_fdcrba_dqdot = 4413


ur5_nodes_id_dqddot= 809

ur5_nodes_fdaba_dtau = 2186

ur5_nodes_fdcrba_dtau = 1695



ur5_nodes_id= 696

ur5_nodes_c= 648

ur5_nodes_m= 1026

ur5_nodes_g= 199

ur5_nodes_fdaba 1948

ur5_nodes_fdcrba= 2354


ur5_nodes_id_dq= 3069

ur5_nodes_c_dq= 3006

ur5_nodes_m_dq = 4095

ur5_nodes_g_dq = 881

ur5_nodes_fdaba_dq = 10189

ur5_nodes_fdcrba_dq = 15194


ur5_nodes_id_dqdot= 2371

ur5_nodes_c_dqdot = 3006

ur5_nodes_fdaba_dqdot = 4605

ur5_nodes_fdcrba_dqdot = 4413


ur5_nodes_id_dqddot= 809

ur5_nodes_fdaba_dtau = 2186

ur5_nodes_fdcrba_dtau = 1695
